### ISSUE TITLE
systray-spec implementation ?

### ISSUE THREAD

**Issue by muhammadsayuti** [sidorares/node-x11#165]
is it possible to implement this api using this library ? 
if you have any example for implementing this using this library i can try to read it, it's better than reading the c code lol,  i'm trying to make some kind of system tray icon panel.

reference:
[https://specifications.freedesktop.org/systemtray-spec/systemtray-spec-0.2.html](https://specifications.freedesktop.org/systemtray-spec/systemtray-spec-0.2.html)

**Comment by sidorares** [comment]
yes, should be doable. Have a look first at https://github.com/santigimeno/node-ewmh as it can have some useful helpers

The systray spec is built on top of XEMBED protocol ( https://standards.freedesktop.org/xembed-spec/xembed-spec-latest.html ) which probably is a good candidate for own library if built with node-x11

**Comment by muhammadsayuti** [comment]
that code is exactly what i read before, but still don't know where to start. I already read some app sample like wingpanel that written in Vala (because i understand C# i think i can understand wingpanel, but i'm wrong).
Have tried to read [cairo-dock systray](https://github.com/Cairo-Dock/cairo-dock-plug-ins/tree/master/systray) but with no luck.
Some people mention in [ubuntu forum](https://ubuntuforums.org/showthread.php?t=2164374) that i need to use dbus. 
But not all example i read using dbus. The closest one i found is [xfce4-panel systray](https://github.com/xfce-mirror/xfce4-panel/blob/master/plugins/systray/systray-manager.c#L302), but still don't know how to implement it using this library.
I think i should send a message first to x11 that i'm going to handle current display, isn't it ?

**Comment by sidorares** [comment]
I haven't read systray docs in detail yet.

Which part do you want to focus first - tray itself or tray icon/applet thing (the one that is embedded)?
Do you want to build example program that sits in the tray or you want to build your own tray?

**Comment by sidorares** [comment]
> Some people mention in ubuntu forum that i need to use dbus.

That's probably completely orthogonal to systray ( and likely unity specific ). I explored a bit of that in https://github.com/sidorares/node-dbusmenu

**Comment by sidorares** [comment]
linking some of my past research on dbus <-> launcher communication:

https://gist.github.com/sidorares/8d0587e384176f1be83d
https://askubuntu.com/questions/530376/association-window-desktop-file-launcher-entry/530431#530431

**Comment by muhammadsayuti** [comment]
i already know how to create tray icon/applet because i already create one using Linux Mint nvidia-prime-applet as my sample.
And now i want to make my own tray panel where i can draw all registered app  that show a tray icon to sit in my tray panel, i just don't know how to get all of those app data, i know that i cun run applet  using .desktop file and put it in /etc/xdg/autostart but what about other gtk app that show a tray icon ?

**Comment by sidorares** [comment]
if you can read C this might be a good start: https://github.com/nsf/bmpanel2/blob/5ad46c872215fb1cd8fdf15ea52808f6991d689a/widget-systray.c

**Comment by muhammadsayuti** [comment]
This is what i get so far 
```javascript

const x11 = require('x11');
const EventEmitter = require('events').EventEmitter;
const {
  Exposure, StructureNotify, ButtonPress, ButtonRelease,
  PointerMotion, EnterWindow, LeaveWindow
} = x11.eventMask;
const set_property = require('x11-prop').set_property;
const promisify = require('util').promisify;

const NET_SYSTEM_TRAY_ORIENTATION_HORZ = 0;
const NET_SYSTEM_TRAY_ORIENTATION_VERT = 1;

class SysTray extends EventEmitter {
  constructor(X, screen) {
    super();
    this.X = X;
    this.screen = screen;
    this.root = screen.root;
    this.sw = {};
    var wid = X.AllocID();
    this.sw.selectionOwner = wid;
    X.CreateWindow(
      wid, this.rootId,        // new window id, parent
      0, 0, 100, 100,   // x, y, w, h
      0, 0, 0, 0,       // border, depth, class, visual
    );
    X.MapWindow(wid);
    var gc = X.AllocID();
    this.sw.gc = gc;
    X.CreateGC(gc, wid);
    this.X.on('event', this.handleEvent.bind(this));
    this.acquireTraySelectionAtom(wid, this.screen.screenNum, (err) => {
      if (err) throw err;
      this.X.InternAtom(wid, "_NET_SYSTEM_TRAY_ORIENTATION", (err, orientatom) => {
        if (err) throw err;
        this.X.InternAtom(wid, "_NET_SYSTEM_TRAY_VISUAL", (err, visualatom) => {
          if (err) throw err;
          this.changeProperty(wid, orientatom, this.X.atoms.INTEGER, 32, [NET_SYSTEM_TRAY_ORIENTATION_HORZ])
            .then(() => {
              return this.changeProperty(wid, visualatom, this.X.atoms.VISUALID, 32, [this.screen.root_visual])
            })
        })
      });
    });
  }
  handleEvent(e) {
    switch (e.name) {
      case 'ClientMessage':
        this.X.GetAtomName(e.type, (err, name) => {
          switch (name) {
            case '_NET_SYSTEM_TRAY_OPCODE':
              this.emit('ActiveWindow', ev.wid);
              break;

            case '_NET_CLOSE_WINDOW':
              this.emit('CloseWindow', ev.wid);
              break;

            case '_NET_CURRENT_DESKTOP':
              this.emit('CurrentDesktop', ev.data[0]);
              break;

            case '_NET_WM_DESKTOP':
              this.emit('Desktop', ev.wid, ev.data[0]);
              break;
          }
        });
        break;
    }
  }
  sendEvent(dest, type, data) {
    this.X.InternAtom(false, "MANAGER", () => {
      var raw = new Buffer(new Array(32));
      raw.writeInt8(33, 0); /* ClientMessage code */
      raw.writeInt8(32, 1); /* Format */
      raw.writeUInt32LE(dest, 4);
      raw.writeUInt32LE(type, 8); /* Message Type */
      raw.writeUInt32LE(data, 12); /* For WM_PROTOCOLS data === Protocol */
      this.X.SendEvent(dest, false, StructureNotify, raw);
    });
  }
  traySelectionOwnerExists(traysel) {
    let old_owner = this.X.GetSelectionOwner(traysel);
    if (old_owner != 0) {
      return 1;
    }
    return 0;
  }
  acquireTraySelectionAtom(wid, screenNum, cb) {
    this.X.InternAtom(false, "_NET_SYSTEM_TRAY_S" + screenNum, (err, tray_selection_atom) => {
      if (err) return cb(err);
      this.X.SetSelectionOwner(wid, tray_selection_atom, cb);
    });
  }

  changeProperty(wid, property, type, format, data, cb) {
    if (!cb && typeof cb !== 'function') {
      const promise = new Promise((resolve, reject) => {
        set_property(this.X, wid, property, type, format, data, resolve);
      });
      return promise;
    } else {
      set_property(this.X, wid, property, type, format, data, cb);
    }
  }
}
module.exports = SysTray;

```

And now i'm stuck on trying to do SendEvent because i don't know all the offset, cab you help with this ?
```C
/* inform other clients that we're here */
XEvent ev;
ev.xclient.type = ClientMessage;
ev.xclient.message_type = XInternAtom(c->dpy, "MANAGER", False);
ev.xclient.display = c->dpy;
ev.xclient.window = c->root;
ev.xclient.format = 32;
ev.xclient.data.l[0] = CurrentTime; /* timestamp */
ev.xclient.data.l[1] = sw->tray_selection_atom; /* manager selection atom */
ev.xclient.data.l[2] = sw->selection_owner; /* the window owning the selection */
ev.xclient.data.l[3] = 0l; /* selection specific data */
ev.xclient.data.l[4] = 0l; /* selection specific data */

XSendEvent(c->dpy, c->root, False, StructureNotifyMask, &ev);
```

**Comment by muhammadsayuti** [comment]
do you know how to convert array of 32bit packed CARDINAL ARGB to base64 or Buffer ??
i'm trying to get the window icon but i don't know how to parse it.

```javascript
getWindowIcon(wid) {
    return new Promise((resolve, reject) => {
      this.X.InternAtom(wid, "_NET_WM_ICON", async (err, atom: number) => {
        if (err) throw err;

        // How to convert this array of 32bit packed CARDINAL ARGB to base64 or Buffer ??
        const icon: any = await GetWindowProperty(this.X, wid, atom, this.X.CARDINAL);

        resolve(icon)
      })
    })
  }
```

**Comment by sidorares** [comment]
how are you going to use it? What format it's currently returned? I thought GetWindowProperty gives you Buffer

**Comment by muhammadsayuti** [comment]
@sidorares edited my code, i'm using x11-prop to get that property cuz i always get BadAtom when using X.GetProperty

**Comment by sidorares** [comment]
according to spec, content of the _NET_WM_ICON is:

> This is an array of 32bit packed CARDINAL ARGB with high byte being A, low byte being B. The first two cardinals are width, height. Data is in rows, left to right and top to bottom.

Which means you need to encode to png after you receive it (if you need png. Other option is to draw this raw image data on canvas ) You can use https://github.com/lukeapage/pngjs for example to do raw -> png encoding

**Comment by sidorares** [comment]
basically it's very similar to what you get from canvas2d.getImageData (maybe order slightly different, for canvas afaik it' RGBA)

**Comment by muhammadsayuti** [comment]
@sidorares what's the diffrence between RGBA and ARGB ? really i don't understand duh. Sorry still new with this kind of thing never play with ARGB before.

**Comment by sidorares** [comment]
CARDINAL is 4 bytes, RGBA - 3 c0lor values + transparency, red / green / blue + "Alpha" - 0-255 values

ARGB is same as RGBA, you just have red red value in the second byte, not the first, green in third etc

**Comment by muhammadsayuti** [comment]
so if i'm using pngjs i just need to parse it like this ? 
```javascript
new PNG({ filterType:4 }).parse(buffer, function(error, data)
{
	console.log(error, data)
});
```

**Comment by sidorares** [comment]
I think .parse is for decoding png data. If you want to encode raw image as png you need to use .pack() - https://github.com/lukeapage/pngjs#pngpack

**Comment by muhammadsayuti** [comment]
@sidorares i change my code to use X.GetProperty and i got the buffer, now where should i put that Buffer in pngjs ??
i've tried to convert it to stream but always get error `There are some read requests waiting on finished stream`

```javascript
const PNG = require('pngjs').PNG;
let Duplex = require('stream').Duplex;
function bufferToStream(buffer) {
  let stream = new Duplex();
  stream.push(buffer);
  stream.push(null);
  return stream;
}
function getWindowIcon(wid) {
    return new Promise((resolve, reject) => {
      this.X.InternAtom(wid, "_NET_WM_ICON", async (err, atom: number) => {
        if (err) throw err;
        this.X.GetProperty(
          0, wid, atom,
          this.X.atoms.CARDINAL, 0 /* 0L */, 10000000 /* ~0L */,
          (err, prop) => {
            const buffer = prop.data;
            var bufferStream = bufferToStream(buffer);
            bufferStream
              .pipe(new PNG({
                filterType: 4
              })).on('parsed', function () {

                this.pack().pipe(fs.createWriteStream('out.png'));
              })
            resolve(buffer.toString('base64'));
          });
        // How to convert this array of 32bit packed CARDINAL ARGB to base64 or Buffer ??
        // const icon: any = await GetWindowProperty(this.X, wid, atom, this.X.CARDINAL);
      })
    })
  }

**Comment by sidorares** [comment]
you don't need to listen for 'parsed' event or stream data to pngjs, this is from decoding example
If you only want to decode you need to do something like this https://github.com/niofis/ymxb/blob/5435602f92829428ac33c7999ebf35a9196ddf6f/wlens/png_test.js

note that first 4 bytes is width and next 4 bytes are height, then in your buffer you have width x height x 4 bytes of ARGB icon data

**Comment by muhammadsayuti** [comment]
@sidorares  is this the right way to read the width and height ?
```javascript
var width = buffer.readUInt32LE(0);
var height = buffer.readUInt32LE(4);
```

**Comment by sidorares** [comment]
hi @muhammadsayuti this is the example of saving icon to png (replace `wid` with your window id)

```js
const x11 = require("x11");
const pngjs = require("pngjs");
const fs = require("fs");

const GetProperty = require("x11-prop").get_property;

const wid = 0x2600001;

x11.createClient(function(err, display) {
  var X = display.client;

  GetProperty(X, wid, "_NET_WM_ICON", "CARDINAL", (err, prop) => {
    const [width, height, ...cardinalPixels] = prop;
    const png = new pngjs.PNG({ width, height });
    const rgbaData = png.data;

    for (var i = 0; i < cardinalPixels.length; ++i) {
      const rgba = Buffer.allocUnsafe(4);
      rgba.writeUInt32LE(cardinalPixels[i]);
      rgbaData[i * 4 + 2] = rgba[0];
      rgbaData[i * 4 + 1] = rgba[1];
      rgbaData[i * 4 + 0] = rgba[2];
      rgbaData[i * 4 + 3] = rgba[3];
    }

    X.close();
    png.pack().pipe(fs.createWriteStream("out.png"));
  });
});
```

**Comment by muhammadsayuti** [comment]
@sidorares i already know how to get the png image, now i'm still confuse about the endianness, you think i should check for endianness too to make it compatible to Big Endian ?

And there is something that i need to know, how can i get the application name instead of window name ? because sometimes window name have some extra detail in it and i don't need that.

Right now i think systray is still on hold because i still can't figure out how to do that but instead i make a dock lol. will try to dig in about systray again later.

**Comment by sidorares** [comment]
@muhammadsayuti I don't have too much time to focus on systray spec, please continue to share with us if you have any success

endianness is defined by client (you):

> The client must send an initial byte of data to identify the byte order to be
employed. The value of the byte must be octal 102 or 154. The value 102 (ASCII
uppercase B) means values are transmitted most significant byte first, and
value 154 (ASCII lowercase l) means values are transmitted least significant
byte first. Except where explicitly noted in the protocol, all 16-bit and
32-bit quantities sent by the client must be transmitted with this byte order,
and all 16-bit and 32-bit quantities returned by the server will be transmitted
with this byte order.

which usually means it's LE. There is a little helper that is called to detect local byte order before connection, you can use same if you really want to be portable

https://github.com/sidorares/node-x11/blob/f040f8bbba7b5be7f1ef1223dde495f92ade3284/lib/handshake.js#L196

**Comment by muhammadsayuti** [comment]
@sidorares i wanna know if SetSelectionOwner callback always get called if it success or error ? i never seee my code being executed inside the callback.

**Comment by sidorares** [comment]
I don't think `SetSelectionOwner` callback is ever called. The way x11 protocol is designed is response is sent only for requests asking for data. Most resource id's are generated on the client and are passed to requests creating those resources so again no need to return anything. Because of this x11 is very optimized for high latency networks
